src/audio_unit/delay.rs

```diff
@@ -1,32 +1,102 @@
+    pub const MIN_DELAY_MS: DelayMs = 0;
+    pub const MAX_DELAY_MS: DelayMs = 10000;
 
-        let ring = RingBuffer::new(delay_num_samples * 2);
+    pub fn new(
+        stream_config: &StreamConfig,
+        midi_message_handler: MidiMessageHandler,
+        delay_ms: DelayMs,
+    ) -> Result<Self> {
+        Self::validate_delay(delay_ms)?;
+
+        // size the ring buffer so that it can accomodate the largest allowed delay
+        let ring = RingBuffer::new(Self::delay_num_samples(stream_config, Self::MAX_DELAY_MS) * 2);
         let (mut producer, consumer) = ring.split();
 
-        ring_buffer::write_empty_samples(&mut producer, delay_num_samples)?;
+        ring_buffer::write_empty_samples(
+            &mut producer,
+            Self::delay_num_samples(stream_config, delay_ms),
+        )?;
+
+        Ok(Self {
+            delay_ms,
+            midi_message_handler,
+            stream_config: stream_config.clone(),
+            producer,
+            consumer,
+        })
+    }
+
+    fn validate_delay(delay_ms: DelayMs) -> Result<()> {
+        if delay_ms > Self::MAX_DELAY_MS {
+            return Err(anyhow!(
+                "Delay must be <= {}, but was {}",
+                Self::MAX_DELAY_MS,
+                delay_ms
+            ));
+        } else {
+            Ok(())
+        }
+    }
+
+    fn delay_num_samples(stream_config: &StreamConfig, delay_ms: u32) -> usize {
+        let delay_num_frames = (delay_ms as f32 / 1_000.0) * stream_config.sample_rate.0 as f32;
+        delay_num_frames as usize * stream_config.channels as usize
+    }
+
+    fn set_delay_ms(&mut self, delay_ms: u32) -> Result<()> {
+        Self::validate_delay(delay_ms)?;
+
+        let old_num_samples = Self::delay_num_samples(&self.stream_config, self.delay_ms);
+        let new_num_samples = Self::delay_num_samples(&self.stream_config, delay_ms);
+
+        if new_num_samples >= old_num_samples {
+            ring_buffer::write_empty_samples(
+                &mut self.producer,
+                new_num_samples - old_num_samples,
+            )?;
+        } else {
+            ring_buffer::read_samples(&mut self.consumer, old_num_samples - new_num_samples)?;
+        }
 
-        Ok(Self { producer, consumer })
+        self.delay_ms = delay_ms;
+
+        Ok(())
     }
 }
 
 impl AudioUnit for Delay {
     fn process(&mut self, input: &[f32], output: &mut [f32]) -> Result<()> {
         ring_buffer::write_frame(&mut self.producer, input)?;
-        let samples: Vec<f32> = ring_buffer::read_frame(&mut self.consumer, output.len())?;
+        let samples: Vec<f32> = ring_buffer::read_samples(&mut self.consumer, output.len())?;
         output.copy_from_slice(&samples);
 
         Ok(())
     }
+
+    fn handle_midi_messages(&mut self, midi_messages: &[MidiMessage<'static>]) -> Result<()> {
+        if let Some(new_delay) = (self.midi_message_handler)(midi_messages) {
```

Comment:
Since the audio unit itself is "dumb", it doesn't seem good (or even possible) for it to know how to adjust itself based on the MIDI input, so the effect that creates it passes in a function telling it how to create a new delay based on the inputs. I'm not sure what I think about this architecture.

Reply:
It seems halfway between exposing an API for something else to control the state mutation vs having it have internal knowledge of how it responds to MIDI messages

Like there's currently an implicit contract where the thing this logic gets "outsourced" to + this have to agree that the return value applies to the delay length state. That doesn't seem like it would scale well to more complex MIDI-controllable API's (eg multiple mutable state values per unit)

Could these guys not even be aware of the existence of MIDI and just expose "knobs" in their public API that then the effect (which is what's aware of MIDI) knows how to turn based on the MIDI messages/mapping? That seems more in keeping with them being "dumb"

Reply:
Yeah, that's what I wanted to do, but I couldn't figure out how. It's complicated because of how the units can only have one owner, so the effects can't really keep a mutable reference to them around. Let's discuss.